### DISCUSSION TITLE
"Fixing" HiDPI issues

### DISCUSSION THREAD

**Discussion by elParaguayo** [qtile/qtile#4781] [Ideas]
I'm working on a solution for users with high resolution screens. My current plan is to allow a `dpi=` argument to `Screen` objects in our configs so this can be configured on a per screen basis. 

I then plan to use that value to scale the following aspects in qtile (tick means I've got it working):
- [x] bar height
- [ ] widget text
- [ ] widget icons
- [ ] bar borders
- [ ] window borders

Questions for people here:
- Is this the right approach? If not, what do you recommend?
- What have I missed from the above list?

**Comment by cr1ogen** [comment]
I change the dpi I am changing the screen scaling through autostart.sh with commands that use wlr-randr, since I use wayland. Maybe this can only be implemented by setting the dependency wlr-randr for wayland and xrandr for X11 and putting the commands in the config.py file
Maybe it's not a good idea, or yes, since I'm not a programmer. But you could see it. Thank you

**Reply by elParaguayo** [reply]
Doesn't screen scaling cause blurring? The solution I'm proposing here ensures that the items I've listed would be rendered at that larger size without the need for scaling but that only covers qtile items. App windows etc wouldn't be impacted.

**Reply by cr1ogen** [reply]
I not see it blurring when i scale the resolution

**Reply by cr1ogen** [reply]
The scaling screen solution, I copy from the wiki of sway

**Reply by cr1ogen** [reply]
![240505_17-35-19](https://github.com/qtile/qtile/assets/9537788/3ce28f71-15b4-4aac-bee7-412211b92dca)

**Comment by tych0** [comment]
We talked about this a bit here: https://github.com/qtile/qtile/issues/4693 ; I think you'll need a way to stabilize which screen is which, given that x11 won't order them in the same way across boots. IMO the best way is to use the serial number, because that's what's printed on the back of the monitor. I have [code to grab the serial number](https://gist.github.com/tych0/a0e2d802b4bde6476997051c576d0ea6#file-monitors-py-L66-L69)  using x11+pyedid, but the edid specification is so small (and we probably only care about the serial number + maybe recommended resolutions) that maybe it's worth just coding our own small implementation.

I think you are on the right track in terms of needing scaling calculations for everything. Two points I would offer: 1. you can compute the DPI based on xrandr outputs (I have code for that as well in the above script), so you don't need DPI (or anything) as a user input unless people want to run things at non-native resolutions, 2. it is useful to be able to run at non native resolutions so that you can drag your mouse across screens smoothly; it is blurry as you say, though. I run one 4k monitor and two non-4ks at 4k for exactly this reason. It doesn't look awful, but I agree it doesn't look great :)

Still, having math to do DPI scaling natively in qtile would be great. I'm happy to look at/test any progress you have.

**Reply by elParaguayo** [reply]
Thanks for the comments.

Agreed on the part about stablising which screen is which. I had a brief chat with @jwijenbergh about that point so it might be worth focusing on that first as it brings other benefits with it too. Once that's working we can layer on the dpi parts.

**Reply by tych0** [reply]
Are there some API calls for edid info on wayland? If so, I can send some generic code to do the actual ordering, and the x11 patches for getting the edid bits. If wayland does its own stable ordering, though, maybe they don't even need this.

**Reply by jwijenbergh** [reply]
Yes, see https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/0.16/include/wlr/types/wlr_output.h?ref_type=heads#L132

**Reply by tych0** [reply]
oh, nice. i can probably even write an untested version of the wayland code with that link. i'll do that soon, thanks!

**Reply by elParaguayo** [reply]
OK. Help me out here. If I adjust the `scale` value in wephyr and `reconfigure_screens`, everything's scaled, windows, bar, borders etc and it looks blurry. I had assumed that's what we're trying to avoid i.e. we can keep scale at 1 but scale up bar, borders etc before rendering. Am I missing something?

**Reply by cr1ogen** [reply]
Shouldn't they deploy wlroots to 0.17 before making these changes?

**Reply by elParaguayo** [reply]
This one's going to take a while anyway and we need something for x11, not just wayland.

**Comment by tych0** [comment]
> OK. Help me out here. If I adjust the scale value in wephyr and reconfigure_screens, everything's scaled, windows, bar, borders etc and it looks blurry. I had assumed that's what we're trying to avoid i.e. we can keep scale at 1 but scale up bar, borders etc before rendering. Am I missing something?

I think you're right that we want the *default* to be scale=1, but I am interested in adjusting the scale down instead of up: I want to render 3x 4k monitors in software, but scale it down by ~0.6 to render on my physical displays, which are really 1920x1200 or whatever. (My practical reason for this is so that my mouse lines up when I drag it across monitors. Maybe there's actually a better way to implement that, though? In which case, perhaps we'd always want scale=1.) I agree that once we have this capability, it doesn't really make sense for anyone to set scale > 1, though. They will definitely prefer rendering at the native resolution.

**Reply by tych0** [reply]
Yeah... https://superuser.com/questions/1119204/move-mouse-proportionally-between-different-resolution-and-dpi-displays looks like it has a few different solutions. Maybe I'll play around with those. I agree it would be much simpler/better to render everything at its preferred native resolution.

**Comment by richcarni** [comment]
I've been testing and experimenting with hidpi support (i.e. fractional scaling) on the Wayland backend - specifically in relation to the bar, since as far as I can tell it's working perfectly fine with (most) apps. I'm wondering if anything discussed here progressed much further? I guess I hope I'm not redoing work that has already been done!

BTW, my setup is a single 27" 4k monitor, for which my preference is 1.5x scaling

My starting point was PR #4625. I fixed some issues with tests and then kept going. I removed the scale factor from config and instead obtain it from wlr_output. And I've modified svg support to fix blurry svg icons.

But issues I still have are:
- Blurry pixmap icons
- Small gaps between widgets

![bar_compare_blur](https://github.com/user-attachments/assets/9b0791bf-b0ed-4dd5-a56f-9dcd4792d3d7)

These are screenshots before and after fixing svg scaling. I think you can see which icons are svgs (BTW, top image wifi is 4 bars!)  You can also see the gaps between widgets - either side of the icons here, but strangely they do move around!

I don't think these are going to be too difficult to solve, but maybe closer to the topic of this discussion, I had some more general thoughts on unifying hidpi support across Wayland and X11...

The crux of the solution in #4625, is to simply scale the context before it is rendered. I believe this works because we're using a recording surface. So text and vector drawings scale without blurring, but not so for pixmaps which are already rasterised

Anyway, it made think why we couldn't do the same for X11? We need to get the scale factor somehow (which I can see was already discussed here). And I can foresee we might need to translate the mouse position for buttons clicks. But I can't see any other issues. However, I definitely may be missing something here!

I am keen to keep progressing things on the Wayland side. I'm super receptive to feedback (I haven't contributed to this codebase before and in many ways am still familiarising myself with it!)
